pkg/lib/cockpit-components-upload.jsx

```diff
@@ -0,0 +1,157 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import cockpit from "cockpit";
+
+const BLOCK_SIZE = 16 * 1024;
+const TOTAL_FRAME_WINDOW = 128;
+/*
+ * Something smart about upload
+ */
+export class UploadHelper {
+    constructor(destination, onProgress) {
+        this.destination = destination;
+        this.progressCallback = onProgress;
+        this.count = TOTAL_FRAME_WINDOW;
+
+        // Promise resolvers
+        this.resolveClose = null;
+        this.resolveFlush = null;
+        this.resolveHandler = null;
+
+        this.channel = cockpit.channel({
+            binary: true,
+            payload: "fsreplace1",
+            path: this.destination,
+            superuser: "try",
```

Comment:
This seems wrong.  It should at least be something left to a `options` argument, if the user really intends for that.

This goes into a bigger conversation about which user `cockpit-files` should be doing operations as when authenticated as root...

Reply:
Agreeed, added options.